plugin/src/main/scala-3/chisel3/internal/plugin/BundlePhase.scala

```diff
@@ -0,0 +1,228 @@
+// SPDX-License-Identifier: Apache-2.0
+
+package chisel3.internal.plugin
+
+import dotty.tools.dotc.report
+import dotty.tools.dotc.ast.tpd
+import dotty.tools.dotc.ast.tpd.TreeOps
+import dotty.tools.dotc.core.Contexts.*
+import dotty.tools.dotc.core.Symbols.*
+import dotty.tools.dotc.core.Names
+import dotty.tools.dotc.core.StdNames.nme
+import dotty.tools.dotc.core.Constants.Constant
+import dotty.tools.dotc.typer.TyperPhase
+import dotty.tools.dotc.plugins.PluginPhase
+import dotty.tools.dotc.core.Types.*
+import dotty.tools.dotc.core.Flags
+import dotty.tools.dotc.core.Decorators.toTermName
+import dotty.tools.dotc.core.Definitions
+
+import scala.collection.mutable
+
+object BundleHelpers {
+  def cloneTypeFull(tree: tpd.Tree)(using Context): tpd.Tree = {
+    val cloneSym =
+      requiredMethod("chisel3.reflect.DataMirror.internal.chiselTypeClone")
+    tpd.Apply(
+      tpd.TypeApply(
+        tpd.ref(cloneSym),
+        List(tpd.TypeTree(tree.tpe))
+      ),
+      List(tree)
+    )
+  }
+
+  def generateAutoCloneType(
+    record:  tpd.TypeDef,
+    thiz:    tpd.This,
+    conArgs: List[List[tpd.Tree]]
+  )(using Context): tpd.DefDef = {
+    val newExpr = tpd.New(record.symbol.typeRef, conArgs.flatten)
+    val recordTpe = requiredClassRef("chisel3.Record")
+    val cloneTypeSym = newSymbol(
+      record.symbol,
+      Names.termName("_cloneTypeImpl"),
+      Flags.Method | Flags.Override | Flags.Protected,
+      MethodType(Nil, Nil, recordTpe)
+    )
+    tpd.DefDef(cloneTypeSym.asTerm, _ => newExpr)
+  }
+
+  // Creates a list of constructor parameter accessors from the
+  // argument value. Returns a Some if a constructor is found, or
+  // None otherwise.
+  def extractConArgs(
+    record: tpd.TypeDef,
+    thiz:   tpd.This
+  )(using Context): Option[List[List[tpd.Tree]]] = {
+    val template = record.rhs.asInstanceOf[tpd.Template]
+    val primaryConstructorOpt = Option(template.constr)
+    val paramAccessors = record.symbol.primaryConstructor.paramSymss.flatten
+
+    if (primaryConstructorOpt.isEmpty) {
+      report.warning("Unable to determine primary constructor!", record.sourcePos)
+      return None
+    }
+
+    val constructor = primaryConstructorOpt.get
+    val paramLookup = paramAccessors.map(sym => sym.name.toString -> sym).toMap
+    val symAccessors = (x: Names.TermName) => record.symbol.asClass.paramAccessors.find(_.name == x)
+
+    if (constructor.symbol.is(Flags.Private)) {
+      val msg = "Private bundle constructors cannot automatically be cloned, try making it package private"
+      report.error(msg, constructor.srcPos)
+      return None
+    }
+
+    constructor.termParamss.map(_.map {
+      case param if symAccessors(param.name).forall(_.is(Flags.Private)) =>
+        report.warning("Private Bundle constructor parameters render the constructor uncloneable")
+        return None
+      case _ =>
+    })
```

Comment:
```suggestion
    constructor.termParamss.foreach(_.foreach {
      case param if symAccessors(param.name).forall(_.is(Flags.Private)) =>
        report.warning("Private Bundle constructor parameters render the constructor uncloneable")
        return None
      case _ =>
    })
```
No return value, but the report also needs to include a source position, at least the constructor if not the exact argument.

Reply:
looks great now
```
[warn] 17 |class ArbiterIO[T <: Data](private val gen: T, val n: Int) extends Bundle {
[warn]    |               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[warn]    |      Private Bundle constructor parameters render the constructor
```